### PR TITLE
API change when using assembled jacobians.

### PR DESCRIPTION
**Pull request by naylor-b**
There are now two options,  'assembled_jac_type' on System and 'assemble_jac' on linear solvers that determine if an assembled jacobian will be used for any given linear solver.  Possible values for 'assembled_jac_type' are 'dense' and 'csc', with 'csc' being the default.  The 'assemble_jac' option on linear solvers is a boolean.

### CODE REVIEW

File: openmdao/components/balance_comp.py [hunk 1]
```diff
@@ -156,8 +156,10 @@ def setup(self):
             self._dscale_drhs = np.ones(n)
 
             ar = np.arange(n)
-            self.declare_partials(of=name, wrt=options['lhs_name'], rows=ar, cols=ar, val=1.0)
-            self.declare_partials(of=name, wrt=options['rhs_name'], rows=ar, cols=ar, val=1.0)
+            self.declare_partials(of=name, wrt=options['lhs_name'], rows=ar, cols=ar,
+                                  val=np.ones(n))
+            self.declare_partials(of=name, wrt=options['rhs_name'], rows=ar, cols=ar,
+                                  val=np.ones(n))
```
[the remark this answers is not in the document]
Comment (naylor-b) [reply openmdao/components/balance_comp.py]: It works either way.  I changed those while debugging and just never changed them back (until now).

File: openmdao/solvers/solver.py [hunk 2]
```diff
@@ -210,6 +212,13 @@ def __init__(self, **kwargs):
 
         self.cite = ""
 
+    def _assembled_jac_solver_iter(self):
+        """
+        Return an empty generator of lin solvers using assembled jacs.
+        """
+        for i in ():
```
[the remark this answers is not in the document]
Comment (naylor-b) [reply openmdao/solvers/solver.py]: Yeah it's an empty generator.  If this is too weird I could replace all of the generators with functions that return lists and I could just add the lists and do a little copying.

File: openmdao/approximation_schemes/complex_step.py [hunk 3]
```diff
@@ -103,7 +103,7 @@ def compute_approximations(self, system, jac=None, deriv_type='partial'):
             being approximated.
         """
         if jac is None:
-            jac = system._jacobian
+            jac = system.jacobian
```
[the remark this answers is not in the document]
Comment (naylor-b) [reply openmdao/approximation_schemes/complex_step.py]: I was using the property to return the assembled_jac if it was there else self._jacobian.  I guess that's not the most obvious behavior.  I can get rid of it.

File: openmdao/core/group.py [hunk 4]
```diff
@@ -1794,19 +1811,18 @@ def _setup_jacobians(self, jacobian=None, recurse=True):
 
         Parameters
         ----------
-        jacobian : <AssembledJacobian> or None
-            The global jacobian to populate for this system.
         recurse : bool
-            Whether to call this method in subsystems.
+            If True, setup jacobians i descendants.
```
[the remark this answers is not in the document]
Comment (naylor-b) [reply openmdao/core/group.py]: fixed

File: openmdao/approximation_schemes/complex_step.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (naylor-b) [reply openmdao/approximation_schemes/complex_step.py]: OK it's gone, along with jacobian.setter.

File: openmdao/core/explicitcomponent.py [hunk 5]
```diff
@@ -2,6 +2,8 @@
 
 from __future__ import division
 
+from copy import deepcopy
```
[the remark this answers is not in the document]
Comment (naylor-b) [reply openmdao/core/explicitcomponent.py]: fixed

File: openmdao/docs/features/core_features/working_with_derivatives/assembled_jacobian.rst [hunk 6]
```diff
@@ -4,22 +4,33 @@
...
+    Assembled Jacobians are especially effective when you have a deeply-nested hierarchy with a
+    large number of components and/or variables. See the
+    :ref:`Theory Manual entry on assembled Jacobians<theory_assembled_vs_matrix_free>` for more
+    details on how to best select which type of Jacobian to use.
 
 
-To use an assembled Jacobian, you set the :code:`jacobian` attribute of the component or group you would like to have assembled.
-There are two kinds of assembled Jacobian you can select from:
+To use an assembled Jacobian, you set the :code:`assemble_jac` option of the linear solver that
+will own it to True.  The type of the assembled jacobian will be determined by the value of
```
[the remark this answers is not in the document]
Comment (kmarsteller) [reply openmdao/docs/features/core_features/working_with_derivatives/assembled_jacobian.rst]: Also, if you're changing it, that instance of "jacobian" probably needs to be capitalized?

**Pull request merged**